prow/tide/tide.go

```diff
@@ -932,12 +932,54 @@ func (c *Controller) accumulateBatch(sp subpool) (successBatch []PullRequest, pe
 	return successBatch, pendingBatch
 }
 
+// prowJobsFromContexts constructs ProwJob objects from all successful presubmit contexts that include a baseSHA.
+// This is needed because otherwise we would always need retesting for results that are older than sinkers
+// max_prowjob_age.
+func prowJobsFromContexts(l *logrus.Entry, ghc githubClient, pr *PullRequest, baseSHA string) ([]prowapi.ProwJob, error) {
+	headContexts, err := headContexts(l, ghc, pr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get head contexts: %w", err)
+	}
+	var passingCurrentContexts []string
+	for _, headContext := range headContexts {
+		if headContext.State != githubql.StatusStateSuccess {
+			continue
+		}
+		if baseSHAForContext := config.BaseSHAFromContextDescription(string(headContext.Description)); baseSHAForContext != "" && baseSHAForContext == baseSHA {
+			passingCurrentContexts = append(passingCurrentContexts, string((headContext.Context)))
+		}
+	}
+
+	var prowjobsFromContexts []prowapi.ProwJob
+	for _, passingCurrentContext := range passingCurrentContexts {
+		prowjobsFromContexts = append(prowjobsFromContexts, prowapi.ProwJob{
+			Spec: prowapi.ProwJobSpec{
+				Context: passingCurrentContext,
+				Refs:    &prowapi.Refs{Pulls: []prowapi.Pull{{Number: int(pr.Number), SHA: string(pr.HeadRefOID)}}},
+				Type:    prowapi.PresubmitJob,
+			},
+			Status: prowapi.ProwJobStatus{
+				State: prowapi.SuccessState,
+			},
+		})
+	}
+
+	return prowjobsFromContexts, nil
+}
+
 // accumulate returns the supplied PRs sorted into three buckets based on their
 // accumulated state across the presubmits.
-func accumulate(presubmits map[int][]config.Presubmit, prs []PullRequest, pjs []prowapi.ProwJob, log *logrus.Entry) (successes, pendings, missings []PullRequest, missingTests map[int][]config.Presubmit) {
+func accumulate(presubmits map[int][]config.Presubmit, prs []PullRequest, pjs []prowapi.ProwJob, log *logrus.Entry, baseSHA string, ghc githubClient) (successes, pendings, missings []PullRequest, missingTests map[int][]config.Presubmit) {
 
 	missingTests = map[int][]config.Presubmit{}
 	for _, pr := range prs {
+
+		if prowjobsFromContexts, err := prowJobsFromContexts(log, ghc, &pr, baseSHA); err != nil {
+			logrus.WithError(err).Error("failed to get prowjobs from contexts")
+		} else {
+			pjs = append(pjs, prowjobsFromContexts...)
```

Comment:
I think appending means the priority of status context is higher than prowjobs, so if status context says pass and prowjobs says fail, tide will still merge the PR. Is there any potential risk here? For example outdated status context?

Reply:
The status context are retrieved for a sha, so they can not be outdated (except for the basesha which is why we are doing the dance here). Tide always tries to find the best result and will use that (i.E. if you have one failure, one success, the job is considered a success) so the order shouldn't be relevant: https://github.com/kubernetes/test-infra/blob/eeb7afb52810b459ddc23b33890e69cbf8efb61f/prow/tide/tide.go#L983

Reply:
I see, this makes sense. Since they are the same sha and baseSha, the order shouldn't matter. Even in worst case, say a flaky job passed then failed, tide will still respect the passed context if the failed job hasn't been reported by crier yet, which is totally fine with me, and I would even consider it a bonus. So SGTM

Reply:
Yeah see like *this* conversation also does not have to happen if we generate all contexts with correct timestamps etc and do filtering after.